File: jas/tools/anchor_point_tool_test.py
"""Anchor Point (Convert) tool tests.

Mirrors jas_dioxus/src/tools/anchor_point_tool.rs `mod tests`. The
meaningful coverage is in the geometry helpers
(convert_corner_to_smooth / convert_smooth_to_corner /
move_path_handle_independent / is_smooth_point), which the tool
sequences.
"""

from __future__ import annotations

from geometry.element import (
    CurveTo, LineTo, MoveTo, Path,
    convert_corner_to_smooth, convert_smooth_to_corner,
    is_smooth_point, move_path_handle_independent,
)


def make_line_path() -> Path:
    return Path(d=(
        MoveTo(0.0, 0.0),
        LineTo(50.0, 0.0),
        LineTo(100.0, 0.0),
    ))


def make_smooth_path() -> Path:
    return Path(d=(
        MoveTo(0.0, 0.0),
        CurveTo(10.0, 20.0, 40.0, 20.0, 50.0, 0.0),
        CurveTo(60.0, -20.0, 90.0, -20.0, 100.0, 0.0),
    ))


def approx_eq(a: float, b: float) -> bool:
    return abs(a - b) < 0.01


def test_corner_point_is_not_smooth():
    pe = make_line_path()
    assert not is_smooth_point(pe.d, 0)
    assert not is_smooth_point(pe.d, 1)
    assert not is_smooth_point(pe.d, 2)


def test_smooth_point_is_smooth():
    pe = make_smooth_path()
    assert is_smooth_point(pe.d, 1)


def test_convert_corner_to_smooth_creates_handles():
    pe = make_line_path()
    result = convert_corner_to_smooth(pe, 1, 50.0, 30.0)
    assert isinstance(result.d[1], CurveTo)
    # Outgoing handle on the next segment is at (50, 30).
    nc = result.d[2]
    assert isinstance(nc, CurveTo)
    assert approx_eq(nc.x1, 50.0)
    assert approx_eq(nc.y1, 30.0)
    # Incoming handle on this segment is reflected: (50, -30).
    cur = result.d[1]
    assert isinstance(cur, CurveTo)
    assert approx_eq(cur.x2, 50.0)
    assert approx_eq(cur.y2, -30.0)


def test_convert_first_anchor_corner_to_smooth():
    pe = make_line_path()
    result = convert_corner_to_smooth(pe, 0, 10.0, 20.0)
    nc = result.d[1]
    assert isinstance(nc, CurveTo)
    assert approx_eq(nc.x1, 10.0)
    assert approx_eq(nc.y1, 20.0)


def test_convert_last_anchor_corner_to_smooth():
    pe = make_line_path()
    result = convert_corner_to_smooth(pe, 2, 100.0, 30.0)
    cur = result.d[2]
    assert isinstance(cur, CurveTo)
    # Reflected of (100, 30) through (100, 0) = (100, -30).
    assert approx_eq(cur.x2, 100.0)
    assert approx_eq(cur.y2, -30.0)
    assert approx_eq(cur.x, 100.0)
    assert approx_eq(cur.y, 0.0)


def test_convert_smooth_to_corner_collapses_handles():
    pe = make_smooth_path()
    result = convert_smooth_to_corner(pe, 1)
    assert not is_smooth_point(result.d, 1)
    cur = result.d[1]
    assert isinstance(cur, CurveTo)
    assert approx_eq(cur.x2, cur.x)
    assert approx_eq(cur.y2, cur.y)
    nc = result.d[2]
    assert isinstance(nc, CurveTo)
    assert approx_eq(nc.x1, 50.0)
    assert approx_eq(nc.y1, 0.0)


def test_independent_handle_move_does_not_reflect():
    pe = make_smooth_path()
    result = move_path_handle_independent(pe, 1, "out", 10.0, 5.0)
    nc = result.d[2]
    assert isinstance(nc, CurveTo)
    assert approx_eq(nc.x1, 70.0)        # 60 + 10
    assert approx_eq(nc.y1, -15.0)       # -20 + 5
    # Incoming handle on cmd[1] is unchanged.
    cur = result.d[1]
    assert isinstance(cur, CurveTo)
    assert approx_eq(cur.x2, 40.0)
    assert approx_eq(cur.y2, 20.0)

File: jas_ocaml/test/tools/anchor_point_tool_test.ml
(** Anchor Point (Convert) tool tests. Mirrors the tests in
    jas_dioxus/src/tools/anchor_point_tool.rs. The meaningful coverage
    lives in the geometry helpers (convert_corner_to_smooth /
    convert_smooth_to_corner / move_path_handle_independent /
    is_smooth_point), which the tool just sequences. *)

open Jas.Element

let make_line_path () = [
  MoveTo (0.0, 0.0);
  LineTo (50.0, 0.0);
  LineTo (100.0, 0.0);
]

let make_smooth_path () = [
  MoveTo (0.0, 0.0);
  CurveTo (10.0, 20.0, 40.0, 20.0, 50.0, 0.0);
  CurveTo (60.0, -20.0, 90.0, -20.0, 100.0, 0.0);
]

let nth d i = List.nth d i

let approx_eq a b = abs_float (a -. b) < 0.01

let () =
  Alcotest.run "Anchor_point_tool" [
    "smooth detection", [
      Alcotest.test_case "corner point is not smooth" `Quick (fun () ->
        let d = make_line_path () in
        assert (not (is_smooth_point d 0));
        assert (not (is_smooth_point d 1));
        assert (not (is_smooth_point d 2)));

      Alcotest.test_case "smooth point is smooth" `Quick (fun () ->
        let d = make_smooth_path () in
        assert (is_smooth_point d 1));
    ];

    "corner to smooth", [
      Alcotest.test_case "convert corner to smooth creates handles" `Quick (fun () ->
        let d = make_line_path () in
        let result = convert_corner_to_smooth d 1 50.0 30.0 in
        (match nth result 1 with
         | CurveTo (_, _, x2, y2, _, _) ->
           assert (approx_eq x2 50.0);
           assert (approx_eq y2 (-30.0))
         | _ -> failwith "expected CurveTo at index 1");
        (match nth result 2 with
         | CurveTo (x1, y1, _, _, _, _) ->
           assert (approx_eq x1 50.0);
           assert (approx_eq y1 30.0)
         | _ -> failwith "expected CurveTo at index 2"));

      Alcotest.test_case "convert first anchor corner to smooth" `Quick (fun () ->
        let d = make_line_path () in
        let result = convert_corner_to_smooth d 0 10.0 20.0 in
        match nth result 1 with
        | CurveTo (x1, y1, _, _, _, _) ->
          assert (approx_eq x1 10.0);
          assert (approx_eq y1 20.0)
        | _ -> failwith "expected CurveTo");

      Alcotest.test_case "convert last anchor corner to smooth" `Quick (fun () ->
        let d = make_line_path () in
        let result = convert_corner_to_smooth d 2 100.0 30.0 in
        match nth result 2 with
        | CurveTo (_, _, x2, y2, x, y) ->
          (* Reflected of (100, 30) through (100, 0) = (100, -30) *)
          assert (approx_eq x2 100.0);
          assert (approx_eq y2 (-30.0));
          assert (approx_eq x 100.0);
          assert (approx_eq y 0.0)
        | _ -> failwith "expected CurveTo");
    ];

    "smooth to corner", [
      Alcotest.test_case "convert smooth to corner collapses handles" `Quick (fun () ->
        let d = make_smooth_path () in
        let result = convert_smooth_to_corner d 1 in
        assert (not (is_smooth_point result 1));
        (match nth result 1 with
         | CurveTo (_, _, x2, y2, x, y) ->
           assert (approx_eq x2 x);
           assert (approx_eq y2 y)
         | _ -> ());
        (match nth result 2 with
         | CurveTo (x1, y1, _, _, _, _) ->
           assert (approx_eq x1 50.0);
           assert (approx_eq y1 0.0)
         | _ -> ()));
    ];

    "independent handle move", [
      Alcotest.test_case "independent handle move does not reflect" `Quick (fun () ->
        let d = make_smooth_path () in
        let result = move_path_handle_independent d 1 "out" 10.0 5.0 in
        (match nth result 2 with
         | CurveTo (x1, y1, _, _, _, _) ->
           assert (approx_eq x1 70.0);    (* 60 + 10 *)
           assert (approx_eq y1 (-15.0))  (* -20 + 5 *)
         | _ -> ());
        (* Incoming handle on cmd[1] is unchanged. *)
        (match nth result 1 with
         | CurveTo (_, _, x2, y2, _, _) ->
           assert (approx_eq x2 40.0);
           assert (approx_eq y2 20.0)
         | _ -> ()));
    ];
  ]
